File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
class Solution:
    # bottom-up DP solution, O(n^2) time and space
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [[0] * (n + 1) for _ in range(n + 1)]

        for i in range(n - 1, -1, -1):
            for j in range(i - 1, -2, -1):
                # not including nums[i] in lis
                lis = dp[i + 1][j + 1]
                
                # including nums[i] in lis
                if j == -1 or nums[j] < nums[i]:
                    lis = max(lis, 1 + dp[i + 1][i + 1])
                
                dp[i][j + 1] = lis

        return dp[0][0]

File: Data Structures & Algorithms/longest-strictly-increasing-or-strictly-decreasing-subarray/submission-3.cpp
class Solution {
public:
    // one-pass solution, O(n) time, O(1) space
    int longestMonotonicSubarray(vector<int>& nums) {
        int increase = 1, decrease = 1;
        int longest = 1;

        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] == nums[i - 1]) {
                increase = 1;
                decrease = 1;
            }
            else if (nums[i] > nums[i - 1]) {
                increase++;
                decrease = 1;
            }
            else {
                increase = 1;
                decrease++;
            }
            longest = max(longest, max(increase, decrease));
        }

        return longest;
    }
};